File: _정답코드/5주차/08_십자가_부분합/solution.py
def solution(numbers, queries):
    n = len(numbers)
    m = len(numbers[0])
    
    rowPrefixSum = [[0] * (m + 1) for _ in range(n + 1)]
    colPrefixSum = [[0] * (m + 1) for _ in range(n + 1)]
    
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            rowPrefixSum[i][j] = rowPrefixSum[i][j - 1] + numbers[i - 1][j - 1]
            colPrefixSum[i][j] = colPrefixSum[i - 1][j] + numbers[i - 1][j - 1]
    
    result = []
    for i in range(len(queries)):
        cx = queries[i][0]
        cy = queries[i][1]
        s = queries[i][2]
        
        horizontalSum = rowPrefixSum[cx][cy + s] - rowPrefixSum[cx][cy - s - 1]
        verticalSum = colPrefixSum[cx + s][cy] - colPrefixSum[cx - s - 1][cy]
        
        result.append(horizontalSum + verticalSum - numbers[cx - 1][cy - 1])
    
    return result

File: _정답코드/5주차/08_십자가_부분합/Solution.java
class Solution {
    public int[] solution(int[][] numbers, int[][] queries) {
        int n = numbers.length;
        int m = numbers[0].length;
        
        int[][] rowPrefixSum = new int[n + 1][m + 1];
        int[][] colPrefixSum = new int[n + 1][m + 1];
        
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                rowPrefixSum[i][j] = rowPrefixSum[i][j - 1] + numbers[i - 1][j - 1];
                colPrefixSum[i][j] = colPrefixSum[i - 1][j] + numbers[i - 1][j - 1];
            }
        }
        
        int[] result = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            int cx = queries[i][0];
            int cy = queries[i][1];
            int s = queries[i][2];
            
            int horizontalSum = rowPrefixSum[cx][cy + s] - rowPrefixSum[cx][cy - s - 1];
            int verticalSum = colPrefixSum[cx + s][cy] - colPrefixSum[cx - s - 1][cy];
            
            result[i] = horizontalSum + verticalSum - numbers[cx - 1][cy - 1];
        }
        
        return result;
    }
}

